EduVPN-redesign/Controllers/ConnectionViewController.swift

```diff
@@ -10,6 +10,10 @@ import AppKit
 import PromiseKit
 import os.log
 
+protocol ConnectionViewControllerDelegate: class {
+    func connectionViewControllerAttemptingToConnect(connectionAttempt: ConnectionAttempt?)
```

Comment:
Delegates should always be pass the caller. So I'd expect something like `func connectionViewController(_ controller: ConnectionViewController, attemptingToConnect connectionAttempt: ConnectionAttempt?)`

Reply:
> Delegates should always be pass the caller.

If the parameter is not used, I don't think we should have the extra parameter just for the sake of convention.

In Apple APIs, they have to consider the case where the same object can be the delegates for multiple instances of the delegating class (for example, one view controller managing multiple table views), so it needs to pass the delegating instance so that the delegate object can figure out which instance the call is for. We don't have that scenario here, so we need not pass the instance. If we have no need for it, we should not do it. The name of the delegate still follows the usual naming convention, so it's still readable without the parameter.

Reply:
It is a convention for a good reason. While you may not need it now, you are now baking the assumption into ConnectionViewController that it won't ever have a sibling. which is something it should not actually be aware off. Sure, this current code works and you'll probably not gonna need it, but still I'd rather see all delegate protocols according to the convention.

Reply:
I agree with Johan on this one.

Delegation is a pattern and in this pattern it is standard practice to pass the caller of the delegate. The delegate has no clue who is calling through the delegated call.

The idea of passing the view controller to the delegate is to maintain the original context (the View Controller) at the location of the delegate implementation.

https://en.wikipedia.org/wiki/Delegation_pattern

Reply:
I also use the delegate pattern for the view model (the delegating object) to tell something to the corresponding view controller (the delegate object). Here we know there can be only one view model for the view controller ever. Do you want the call to include the view model instance in this case?

Reply:
I pushed commits separately for view controllers, services and view models being delegating objects. Let me know what you think. In case you think it doesn't apply for any of them, we can revert the corresponding commit.

If you think this looks good as it is, I'll merge it in.

Reply:
I'm proceeding to merge this PR. In case you think there's a problem here, please let me know.